unixfs/src/dir/builder/buffered.rs

```diff
@@ -0,0 +1,464 @@
+    /// completion while storing the created blocks.
+    ///
+    /// Returned `PostOrderIterator` will use the given `full_path` and `block_buffer` to store
+    /// it's data during the walk. `PostOrderIterator` implements `Iterator` while also allowing
+    /// borrowed access via `next_borrowed`.
+    pub fn build<'a>(
+        self,
+        full_path: &'a mut String,
+        block_buffer: &'a mut Vec<u8>,
+    ) -> PostOrderIterator<'a> {
+        PostOrderIterator::new(
+            Visited::Descent {
+                node: self.root_builder,
+                name: None,
+                depth: 0,
+            },
+            full_path,
+            block_buffer,
+            self.opts,
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        super::OwnedTreeNode, BufferingTreeBuilder, Metadata, TreeBuildingFailed, TreeOptions,
+    };
+    use cid::Cid;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn some_directories() {
+        let mut builder = BufferingTreeBuilder::default();
+
+        // foobar\n
+        let five_block_foobar =
+            Cid::try_from("QmRJHYTNvC3hmd9gJQARxLR1QMEincccBV53bBw524yyq6").unwrap();
+
+        builder
+            .put_file("a/b/c/d/e/f/g.txt", five_block_foobar.clone(), 221)
+            .unwrap();
+        builder
+            .put_file("a/b/c/d/e/h.txt", five_block_foobar.clone(), 221)
+            .unwrap();
+        builder
+            .put_file("a/b/c/d/e/i.txt", five_block_foobar, 221)
+            .unwrap();
+
+        let mut full_path = String::new();
+        let mut buffer = Vec::new();
+
+        let iter = builder.build(&mut full_path, &mut buffer);
+        let mut actual = iter
+            .map(|res| res.map(|n| (n.path, n.cid, n.block)))
+            .collect::<Result<Vec<_>, _>>()
+            .unwrap();
+
+        let mut expected = vec![
+            (
+                "a/b/c/d/e/f",
+                "Qmbgf44ztW9wLcGNRNYGinGQB6SQDQtbHVbkM5MrWms698",
+            ),
+            (
+                "a/b/c/d/e",
+                "Qma1hCr3CuPRAq2Gw4DCNMqsi42Bjs4Bt1MGSS57kNh144",
+            ),
+            ("a/b/c/d", "QmUqaYatcJqiSFdykHXGh4Nog1eMSfDJBeYzcG67KV5Ri4"),
+            ("a/b/c", "QmYwaNBaGpDCNN9XpHmjxVPHmEXZMw9KDY3uikE2UU5fVB"),
+            ("a/b", "QmeAzCPig4o4gBLh2LvP96Sr8MUBrsu2Scw9MTq1EvTDhY"),
+            ("a", "QmSTUFaPwJW8xD4KNRLLQRqVTYtYC29xuhYTJoYPWdzvKp"),
+        ];
+
+        // hopefully this way the errors will be easier to hunt down
+
+        actual.reverse();
+        expected.reverse();
+
+        while let Some(actual) = actual.pop() {
+            let expected = expected.pop().expect("size mismatch");
+            assert_eq!(actual.0, expected.0);
+            assert_eq!(
+                actual.1.to_string(),
+                expected.1,
+                "{:?}: {:?}",
+                actual.0,
+                Hex(&actual.2)
+            );
+        }
+    }
+
+    struct Hex<'a>(&'a [u8]);
```

Comment:
haven't we implemented this bad boi before? it might be a good idea to put utilities like this in a single place accessible to tests, examples etc. A good `FirstIssue` candidate IMO.

Reply:
I always write this again and again... But I've also removed them a lot. I refactored it inside a `fn verify_results(expected, actual)` which now has the shared parts between the two test cases (some_directories and dir_with_cidv1_link).